homeassistant/components/azure_event_hub/__init__.py

```diff
@@ -1,89 +1,240 @@
+                item = self.queue.get(timeout=timeout)
+                dequeue_count += 1
+                if item is None:
+                    self.shutdown = True
+                else:
+                    timestamp, event = item
+                    age = time.monotonic() - timestamp
+                    try:
+                        if age < QUEUE_BACKLOG_SECONDS:
+                            event_data = self.event_to_filtered_event_data(event)
+                            if event_data:
+                                _LOGGER.debug(
+                                    "      Event data json: %s", str(event_data)
+                                )
+                                event_data_batch.add(event_data)
+                        else:
+                            dropped += 1
+                    except ValueError:
+                        can_add = False  # EventDataBatch object reaches max_size.
+        except queue.Empty:
+            pass
+
+        if dropped:
+            _LOGGER.warning("    Catching up, dropped %d old events", dropped)
+        return event_data_batch, dequeue_count
+
+    def send(self):
+        """Write preprocessed events to eventhub, with retry."""
+        additional_args = {"logging_enable": False}
+        if self._conn_str_client:
+            client = EventHubProducerClient.from_connection_string(
+                **self._client_args, **additional_args
+            )
+        else:
+            client = EventHubProducerClient(**self._client_args, **additional_args)
+        data_batch, dequeue_count = self.fill_batch(client.create_batch(max_size=10000))
+        _LOGGER.info(
+            "Sent event count %s, out of %s events in the queue.",
+            str(len(data_batch)),
+            str(dequeue_count),
         )
-        await async_sender.send(event_data)
 
-    async def async_shutdown(event: Event):
-        """Shut down the client."""
-        await client.stop_async()
-
-    hass.bus.async_listen(EVENT_STATE_CHANGED, async_send_to_event_hub)
-    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, async_shutdown)
-
-    return True
+        try:
+            if len(data_batch) > 0:
+                with client:
+                    client.send_batch(data_batch)
+        except EventHubError as exc:
+            _LOGGER.error("Error in sending events to Event Hub: %s", exc)
+        finally:
+            client.close()
+            for _ in range(dequeue_count):
+                self.queue.task_done()
+
+    def run(self):
+        """Process incoming events."""
+
+        @callback
+        def register():
+            """Post connection initialize."""
+            self.async_ready.set_result(True)
+            _LOGGER.debug("Ready to go!")
+
+            def shutdown(event):
+                """Shut down the AEH Thread."""
+                self.queue.put(None)
+                self.join()
+
+            self.hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, shutdown)
```

Comment:
A function in a function in a function. Let make us that flat and register this on top level like in the setup.